File: javascript/src/agent/setup/runner.test.ts
// Tests for the in-memory runner drive helpers.
import { BaseAgent, createEvent, createEventActions, type Event, LlmAgent } from '@google/adk';
import { describe, expect, it } from 'vitest';

import { FakeLlm } from '../../testutil/fakes';
import { drive, driveCollectState, driveText, newRunner } from './runner';

// A minimal custom agent that emits two state deltas.
class Writer extends BaseAgent {
  protected override async *runAsyncImpl(): AsyncGenerator<Event, void> {
    yield createEvent({ author: this.name, actions: createEventActions({ stateDelta: { a: 1 } }) });
    yield createEvent({ author: this.name, actions: createEventActions({ stateDelta: { b: 2 } }) });
  }
  protected override async *runLiveImpl(): AsyncGenerator<Event, void> {
    // not used
  }
}

describe('runner', () => {
  it('driveText returns the agent final text', async () => {
    const agent = new LlmAgent({ name: 'echo', model: new FakeLlm('final answer'), instruction: 'x' });
    const runner = newRunner('t', agent);
    const out = await driveText(runner, 'u', 's1', 'hello');
    expect(out).toContain('final answer');
  });

  it('drive drains events without throwing', async () => {
    const agent = new LlmAgent({ name: 'echo2', model: new FakeLlm('ok'), instruction: 'x' });
    const runner = newRunner('t', agent);
    await expect(drive(runner, 'u', 's2', 'hi')).resolves.toBeUndefined();
  });

  it('driveCollectState accumulates state deltas', async () => {
    const runner = newRunner('t', new Writer({ name: 'w' }));
    const state = await driveCollectState(runner, 'u', 's3', 'go');
    expect(state).toEqual({ a: 1, b: 2 });
  });
});

File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/agent/setup/RunnerTest.kt
package io.github.jkjamies.automationagent.agent.setup

import com.google.adk.kt.agents.BaseAgent
import com.google.adk.kt.agents.InvocationContext
import com.google.adk.kt.events.Event
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/** A minimal custom agent that emits a single text event with a state delta. */
private class EchoAgent : BaseAgent(name = "echo") {
    override fun runAsyncImpl(context: InvocationContext): Flow<Event> = flow {
        emit(textEvent("echo", "hello", mapOf("k" to "v")))
    }
}

/** An agent whose run throws, to prove the driver propagates the failure. */
private class BoomAgent : BaseAgent(name = "boom") {
    override fun runAsyncImpl(context: InvocationContext): Flow<Event> = flow {
        throw RuntimeException("kaboom")
    }
}

class RunnerTest : BehaviorSpec({
    Given("an echo agent behind an in-memory runner") {
        When("driving it") {
            Then("the run completes without error") {
                val r = newRunner("test-app", EchoAgent())
                drive(r, "u", "s", "go")
            }
        }
        When("collecting state") {
            Then("the emitted state delta is accumulated") {
                val r = newRunner("test-app", EchoAgent())
                driveCollectState(r, "u", "s", "go")["k"] shouldBe "v"
            }
        }
    }

    Given("an agent that errors") {
        When("driving it") {
            Then("the error propagates") {
                val r = newRunner("test-app", BoomAgent())
                shouldThrow<RuntimeException> { drive(r, "u", "s", "go") }
            }
        }
    }
})
